src/videoCache.js

```diff
@@ -0,0 +1,213 @@
+      }
+
+      callback(null, ids);
+    },
+    error: function(statusText, responseBody) {
+      callback(new Error('Error storing video ad in the cache: ' + statusText + ': ' + JSON.stringify(responseBody)), []);
+    }
+  }
+}
+
+/**
+ * If the given bid is for a Video ad, generate a unique ID and cache it somewhere server-side.
+ *
+ * @param {CacheableBid[]} bids A list of bid objects which should be cached.
+ * @param {videoCacheStoreCallback} [callback] An optional callback which should be executed after
+ *   the data has been stored in the cache.
+ */
+export function store(bids, callback) {
+  const requestData = {
+    puts: bids.map(toStorageRequest)
+  };
+
+  ajax(PUT_URL, shimStorageCallback(callback), JSON.stringify(requestData), {
+    contentType: 'text/plain',
+    withCredentials: true
+  });
+}
+
+/**
+ * @callback videoCacheRetrieveCallback
+ *
+ * @param {Error} [error] The error, if one occurred.
+ * @param {string} A VAST XML string. If error is defined, this can still expect an "empty"
+ *   VAST URL response.
+ */
+
+/**
+ * Takes the user-defined callback, and wraps it into our ajax function's API.
+ *
+ * This also needs to work around the fact that the cache server returns a 200 on "data not found"...
+ * which this module's API signals as an error.
+ *
+ * @param {videoCacheRetrieveCallback} callback
+ */
+function shimRetrievalCallback(callback) {
+  return {
+    success: function(responseBody) {
+      // If the cache didn't have our data, it returns JSON with an error property.
+      let errorMsg;
+      try {
+        errorMsg = JSON.parse(responseBody).error;
+      }
+      // But if it's not of this form, then we got a cache hit.
+      catch (e) {
+        callback(null, responseBody);
+        return;
+      }
+
+      callback(new Error(errorMsg), EMPTY_VAST_RESPONSE);
+    },
+
+    error: function(statusText, responseBody) {
+      const err = new Error('Error retrieving ad from the cache. ' + statusText + ': ' + responseBody);
+      callback(err, EMPTY_VAST_RESPONSE);
+    }
+  }
+}
+
+/**
+ * Make a server call for the given cacheId, and execute the callback function when they get back to us.
+ *
+ * @param {string} cacheId The ID of the video whose content should be retrieved.
+ * @param {videoCacheRetrieveCallback} callback A callback function which should be run once the
+ *   VAST content has been fetched from the cache.
+ */
+function retrieveFromServer(cacheId, callback) {
+  if (typeof cacheId !== 'string') {
+    callback(new Error('The cacheId must be a string.'), EMPTY_VAST_RESPONSE);
+    return;
+  }
+  ajax(tryAppendQueryString(GET_URL, 'uuid', cacheId).slice(0, -1), shimRetrievalCallback(callback));
+}
+
+/**
+ * Memoization function which prevents excessive server calls. This stores all the
+ * [cacheId, cacheValue] pairs locally so that we don't go to the network again if
+ * the publisher requests the same VAST content more than once.
+ */
+function memoize(retrievalFunction) {
```

Comment:
It's used right below it, in the exported function :).

Per the docs, the goal here is to prevent multiple server calls.

If we successfully fetched a thing from the server, we can store it in page memory for the next time.

If the server fetch failed, we want to retry the call (might have just been a network problem).

It might have some more general uses, but... better private than public and only used once, imo. If refactoring, it's much easier to make more visible rather than less.